Refactor

aekong opened the pull request:
Review site here: http://ci-66-sfdesignsystem.pantheonsite.io

Latest redesign deprecates a lot of the currently existing patterns, so using this opportunity to remove a bunch of stuff, most of it was a straight port from pattern lab out of sf.gov.  Also, trying to make this pattern library more generic (not twig/drupal specific), so going back to handlebars.

Components are in `src/components` and each component has it's own folder with the necessary files.  The following being most common:
- xyz.hbs
- _xyz.scss
- xyz.config.yml
- xyz.js

This PR isn't ready to merge into master yet.  Its intention to get another set of eyes and to start a conversation on whether or not this is a good direction forward.

There's still a bunch of TODO's:
- remove old folder structure (`/components`, `/docs`, `/gulptasks`, `public/css`, `public/fonts`, etc)
- refactor `gulpfile.babel.js` to remove tasks that are no longer relevant
- add tests
- probably other stuff i'm not thinking of

aekong committed:
refactor gulp and config for new structure

aekong committed:
minimum components for review

aekong committed:
fix gulp export task

aekong commented:
Latest redesign deprecates a lot of the currently existing patterns, so using this opportunity to remove a bunch of stuff, most of it was a straight port from pattern lab out of sf.gov.  Also, trying to make this pattern library more generic (not twig/drupal specific), so going back to handlebars.

Components are in `src/components` and each component has it's own folder with the necessary files.  The following being most common:
- xyz.hbs
- _xyz.scss
- xyz.config.yml
- xyz.js

This PR isn't ready to merge into master yet.  Its intention to get another set of eyes and to start a conversation on whether or not this is a good direction forward.

There's still a bunch of TODO's:
- remove old folder structure (`/components`, `/docs`, `/gulptasks`, `public/css`, `public/fonts`, etc)
- refactor `gulpfile.babel.js` to remove tasks that are no longer relevant
- add tests
- probably other stuff i'm not thinking of

jrubenoff commented on .editorconfig:
```diff
@@ -5,4 +5,5 @@ charset = utf-8
-indent_size = 2
\ No newline at end of file
+indent_size = 2
+editor.detectIndentation = false
```
What does this do? Can't find docs for this on editorconfig's site.

aekong replied:
😬 that's a vs code config, sorry

jrubenoff commented on fractal.js:
```diff
@@ -7,22 +7,26 @@ const path = require('path');
 
 /* Set the title of the project */
-fractal.set('project.title', 'SFGOV Pattern Library');
+fractal.set('project.title', 'San Francisco Digital Services Design System');
```
Maybe just "San Francisco Design System?" Ideally other agencies are using it...

jrubenoff commented on src/components/00-design-tokens/colors.hbs:
```diff
@@ -0,0 +1,8 @@
+<ul>
+  {{#each colorset}}
+    <li style="display:block;float:left;margin:0 10px 10px 0;text-align:center;">
```
Can you move all inline styles to `pattern-scaffolding.scss`?

jrubenoff reviewed (changes requested):
Left a few comments inline.

Should I comment on CSS architecture, or is it important that the SF.gov and design system mixins stay in sync?

aekong commented:
Not super important.  It'd be nice to keep variable and mixin names the same, but if there's a good reason for changes (like better practices), totally happy to make those changes.  Otherwise, css architecture is fair game, have at it!

jrubenoff commented on src/components/00-design-tokens/_breakpoints.scss:
```diff
@@ -0,0 +1,5 @@
+$bp-tiny: 375px;
```
Breakpoints should be `rem`s so that they work even when:

- The user has an enlarged default font size
- The zoom level is over 100%
- We increase the `font-size` property in the `html` tag

aekong replied:
these breakpoints are coming from figma: https://www.figma.com/file/YkshTKA1RuQEcIju7cblRR4B/Grids?node-id=0%3A1.  When I convert these to rem's (html font size is 17px) the layout breaks on the incorrect pixel width.

For example, according to the figma file, the a component goes to one column layout at 769px.  When I convert this to rem (45.24rem), the actual pixel break happens at ~732px.

Is this a common scenario?

jrubenoff commented on src/components/00-design-tokens/_breakpoints.scss:
```diff
@@ -0,0 +1,5 @@
+$bp-small: 700px;
+$bp-narrow: 769px;
+$bp-medium: 1024px;
+$bp-wide: 1280px;
```
If we want to namespace these, I feel like `mq-` (media query) might be a more intuitive prefix?

jrubenoff commented on src/components/00-design-tokens/_colors.scss:
```diff
@@ -0,0 +1,34 @@
+$c-white: #fff;
```
@aekong Do you feel like these need to be namespaced? I'm not sure we'll ever be in danger of having two conflicting `$white`s or `$blue-1`s, but maybe you feel differently

jrubenoff commented on src/components/00-design-tokens/_mixins.scss:
```diff
@@ -0,0 +1,11 @@
+@mixin media($bp) {
```
Seems like the sole purpose of these mixins is to save you from typing a few characters. I feel like we should remove them? They do more to obfuscate the code than make us more efficient

jrubenoff commented on src/components/00-design-tokens/_typography-mixins.scss:
```diff
@@ -0,0 +1,37 @@
+@mixin antialiasing($type:antialiased) {
+  /* antialiased, none, subpixel-antialiased*/
```
We need to add some `@if` statements here to incorporate Mozilla's different syntax for this property. (e.g. `-moz-osx-font-smoothing: grayscale;`)

jrubenoff commented on src/components/00-design-tokens/_typography-mixins.scss:
```diff
@@ -0,0 +1,37 @@
+  font-family: "Noto Sans TC", sans-serif;
+}
+
+@mixin fs-title-5 {
```
Not sure these `fs-` mixins benefit from being namespaced either?

Also, we should convert `px` units to `rem`s, for the same reasons mentioned above

jrubenoff commented on src/components/03-containers/three-column-container/_three-column-container.scss:
```diff
@@ -0,0 +1,26 @@
+  }
+  > div {
+    width: 100%;
+    margin-bottom: 10px;
```
Since most of our components have `10px` multiples for spacing and sizing, maybe we should turn that into a `$rhythm` variable (using `rem`s) to embed that into the system?

Examples:

`margin-bottom: $rhythm;`
`padding: $rhythm * 2;`

jrubenoff commented on src/components/03-containers/three-column-container/_three-column-container.scss:
```diff
@@ -0,0 +1,26 @@
+      display: flex;
+      align-items: stretch;
+      &:nth-child(3n+3) {
+        > * {
```
Would it be possible to convert these selectors into classes? Pretty high specificity here.

aekong replied:
i think we talked about this in person, but i forgot where we landed.  The reason i did it here in css was so that it wouldn't need to add a `third-child` class in the markup, which would require some calculation

jrubenoff replied:
To sum up [our Slack convo:](https://sfdigitalservices.slack.com/archives/CEERV5DCG/p1573175275001200?thread_ts=1573170435.000100&cid=CEERV5DCG) it's fine to leave this selector as-is.

Might be a bit more readable if you combined lines 15 and 16, though.

jrubenoff commented:
@aekong Alright, those are all my comments.

aekong committed:
remove vs code specific config

aekong committed:
remove digital services

aekong committed:
remove public folder, now generated

aekong committed:
ignore public folder

aekong committed:
udpate design tokens to use rem instead of px, remove some namespacing

aekong committed:
update component css to use new var names

aekong committed:
imports

aekong committed:
pattern scaffolding for design system static site

aekong committed:
media queries use browser default font size (16px) to calculate rem

aekong committed:
remove media query mixin

aekong committed:
account for moz prefix for antialiasing mixin

aekong committed:
add npm build step

aekong committed:
npm update

Commit:
Update antialiasing mixin

- “font-smoothing” isn’t a real property
- “font-smooth” isn’t recommended on production sites https://developer.mozilla.org/en-US/docs/Web/CSS/font-smooth

jrubenoff reviewed (approved):
I made a small update to the `antialiasing` mixin.

Feel free to merge this in if that looks good to you!

aekong reviewed (commented):
👍
